### PR TITLE
Add ActivitySource.StartActivity overload and support CallerMemberNameAttribute for name parameter

### LINKED ISSUE #43483: Provide default activity name value from [CallerMemberName] in ActivitySource.StartActivity
The improvements up to [part 2](accepted/2020/diagnostics/activity-improvements-2.md) have made things substantially more usable for the Activity API. This is a small proposed improvement to introduce a sensible default activity name.

> Updated proposal based on [feedback](https://github.com/dotnet/runtime/issues/43483#issuecomment-713199370)

## ActivitySource.StartActivity

We currently require an activity `name` to be provided as the only required argument in `ActivitySource.StartActivity` overloads. For scenarios where the caller wants to match the activity with the calling method name, providing that as a default value would make calling code less repetitive.

## ***API  Proposal***

```c#
    public Activity? StartActivity([CallerMemberName] string name = "", ActivityKind kind = ActivityKind.Internal)
            => StartActivity(name, kind, default, null, null, null, default);
```

Additionally, for cases where users want to get the caller member name alongside other parameters, add the following overload too:

## ***API  Proposal***

```c#
public System.Diagnostics.Activity StartActivity (System.Diagnostics.ActivityKind kind, 
                                                  System.Diagnostics.ActivityContext parentContext = default,
                                                  System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string,object>>? tags = default, 
                                                  System.Collections.Generic.IEnumerable<System.Diagnostics.ActivityLink> links = default, 
                                                  DateTimeOffset startTime = default,
                                                  [CallerMemberName] string name = "");
```

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes #43483

### CODE REVIEW

**Comment by tarekgh** [comment]
@noahfalk could you please help reviewing this one? Thanks!

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs [hunk 1]
```diff
@@ -135,9 +135,10 @@ public sealed class ActivitySource : IDisposable
...
         public System.Diagnostics.Activity? StartActivity(string name, System.Diagnostics.ActivityKind kind, string parentId, System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, object?>>? tags = null, System.Collections.Generic.IEnumerable<System.Diagnostics.ActivityLink>? links = null, System.DateTimeOffset startTime = default) { throw null; }
+        public System.Diagnostics.Activity? StartActivity(System.Diagnostics.ActivityKind kind, System.Diagnostics.ActivityContext parentContext = default, System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string, object?>>? tags = null, System.Collections.Generic.IEnumerable<System.Diagnostics.ActivityLink>? links = null, DateTimeOffset startTime = default, [System.Runtime.CompilerServices.CallerMemberName] string name = "") { throw null; }
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:141]: It's a little strange to me that the order of arguments differs between various overloads, but if this is what was agreed on at the API review, ok.

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:141]: Yes that is what have agreed on in the review. But I don't mind moving the `name` parameter to be the second parameter if this make it better. 

@noahfalk @terrajobst what you think?

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 2]
```diff
@@ -125,6 +125,36 @@ public void TestActivityWithListenerActivityCreateAndAllDataRequested()
...
+                {
+                    using ActivityListener listener = new ActivityListener();
+                    listener.ShouldListenTo = (activitySource) => object.ReferenceEquals(aSource, activitySource);
+                    listener.Sample = (ref ActivityCreationOptions<ActivityContext> activityOptions) => ActivitySamplingResult.AllDataAndRecorded;
+
+                    ActivitySource.AddActivityListener(listener);
+
+                    string methodName = MethodBase.GetCurrentMethod().Name;
+
+                    using (Activity activity = aSource.StartActivity()) // passing default name should trigger CallerMemberName attribute.
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:142]: Do we need to test that things like tags, links, startTime, etc. are appropriately propagated?

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs (hunk 2 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:142]: We already testing those with the other overloads. I can add some more validations for this overload.

**Review by noahfalk — APPROVED** [review]
LGTM!

**Pull request merged**